# Project 2 Lab notebook

Fiippov M. @pseudocephalus

 _To run code directly from notebook one should place it in project directory with `/data` subfolder in it._

Reference genome: https://www.ncbi.nlm.nih.gov/nuccore/KF848938.1?report=fasta


Sequencing data: http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR170/001/SRR1705851/


Place both in `/data`, unarchive and rename

In [ ]:
! gunzip  data/*.gz
! mv data/*.fastq data/reads.fastq
! mv data/*.fasta data/ref.fasta

In [ ]:
! bwa index data/ref.fasta

In [ ]:
! bwa mem data/ref.fasta data/reads.fastq | samtools view -S -b - | samtools sort - > data/sorted.bam

In [ ]:
! samtools index data/sorted.bam

In [1]:
! samtools flagstat data/sorted.bam

361349 + 0 in total (QC-passed reads + QC-failed reads)
358265 + 0 primary
0 + 0 secondary
3084 + 0 supplementary
0 + 0 duplicates
0 + 0 primary duplicates
361116 + 0 mapped (99.94% : N/A)
358032 + 0 primary mapped (99.93% : N/A)
0 + 0 paired in sequencing
0 + 0 read1
0 + 0 read2
0 + 0 properly paired (N/A : N/A)
0 + 0 with itself and mate mapped
0 + 0 singletons (N/A : N/A)
0 + 0 with mate mapped to a different chr
0 + 0 with mate mapped to a different chr (mapQ>=5)


Reference genome has length of 1665 bp and sequencing has 358265 reads with average length of 93 bp, so estimated coverage is 

358265 * 93 / 1665 = 20011

In [29]:
! samtools mpileup -d 21000 -f data/ref.fasta data/sorted.bam > data/reads.mpileup

[mpileup] 1 samples in 1 input files


In [3]:
! java -jar VarScan.v2.4.4.jar mpileup2snp data/reads.mpileup --min-var-freq 0.95 --variants --output-vcf 1 > data/VarScan_results.vcf


Only SNPs will be reported
Min coverage:	8
Min reads2:	2
Min var freq:	0.95
Min avg qual:	15
P-value thresh:	0.01
Reading input from data/reads.mpileup
1665 bases in pileup file
5 variant positions (5 SNP, 0 indel)
0 were failed by the strand-filter
5 variant positions reported (5 SNP, 0 indel)


In [19]:
! cat data/VarScan_results.vcf | awk 'NR>24 {print $1, $2, $4, $5}'

KF848938.1 72 A G
KF848938.1 117 C T
KF848938.1 165 T C
KF848938.1 307 C T
KF848938.1 389 T C
KF848938.1 722 A G
KF848938.1 774 T C
KF848938.1 802 A G
KF848938.1 913 T C
KF848938.1 915 T C
KF848938.1 999 C T
KF848938.1 1086 A G
KF848938.1 1213 A G
KF848938.1 1260 A C
KF848938.1 1280 T C
KF848938.1 1458 T C


Inspection in IGV showed that they are all silent

In [30]:
! java -jar VarScan.v2.4.4.jar mpileup2snp data/reads.mpileup --min-var-freq 0.001 --variants --output-vcf 1 > data/VarScan_results.vcf


Only SNPs will be reported
Min coverage:	8
Min reads2:	2
Min var freq:	0.001
Min avg qual:	15
P-value thresh:	0.01
Reading input from data/reads.mpileup
1665 bases in pileup file
23 variant positions (21 SNP, 2 indel)
0 were failed by the strand-filter
21 variant positions reported (21 SNP, 0 indel)


In [31]:
! cat data/VarScan_results.vcf | awk 'NR>24 {print $2, $4, $5 ,$10}'

72 A G 1/1:255:16832:16794:6:16787:99.96%:0E0:35:36:4:2:10898:5889
117 C T 1/1:255:20768:20663:36:20625:99.82%:0E0:35:37:27:9:13462:7163
254 A G 0/1:20:35781:35626:35562:59:0.17%:8.5683E-3:36:36:23919:11643:37:22
276 A G 0/1:24:37022:36965:36892:64:0.17%:3.3004E-3:37:35:22579:14313:30:34
307 C T 0/1:255:37506:37386:37029:351:0.94%:6.9068E-66:36:35:22400:14629:184:167
340 T C 0/1:23:37973:37793:37723:64:0.17%:4.6441E-3:37:36:23413:14310:40:24
389 T C 0/1:40:31960:31763:31690:69:0.22%:9.0525E-5:37:36:16067:15623:44:25
691 A G 0/1:23:39503:39319:39245:67:0.17%:4.1867E-3:37:34:21313:17932:23:44
722 A G 0/1:36:38090:38036:37954:76:0.2%:2.353E-4:37:36:21003:16951:39:37
744 A G 0/1:23:38443:38380:38306:65:0.17%:5.0068E-3:37:33:20859:17447:35:30
774 T C 1/1:255:38397:38191:8:38177:99.96%:0E0:32:37:7:1:19806:18371
802 A G 0/1:59:44236:44106:44001:101:0.23%:1.2224E-6:37:35:20575:23426:32:69
859 A G 0/1:25:35895:35721:35655:63:0.18%:3.0291E-3:37:36:14499:21156:27:36
915 T C 0/1:30:37022:36921:368

16 SNPs was detected with such low min freq with average frequency below 1%. There are two most frequent of rare vaeriants with frequencies of **0.92%** and **0.86%**

Download reference sequencing data

SRR1705858: ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR170/008/SRR1705858/SRR1705858.fastq.gz 

SRR1705859: ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR170/009/SRR1705859/SRR1705859.fastq.gz 

SRR1705860: ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR170/000/SRR1705860/SRR1705860.fastq.gz 

unarchive them in `/data`


In [7]:
! fastqc data/SRR*

SRR1705858: 256586 reads

SRR1705859: 233327 reads

SRR1705860: 249964 reads

Average 246626 reads, coverage = **13775**

In [10]:
! bwa mem data/ref.fasta data/SRR1705858.fastq | samtools view -S -b - | samtools sort - > data/sorted58.bam
! samtools index data/sorted58.bam

In [11]:
! bwa mem data/ref.fasta data/SRR1705859.fastq | samtools view -S -b - | samtools sort - > data/sorted59.bam
! samtools index data/sorted59.bam

[M::bwa_idx_load_from_disk] read 0 ALT contigs
[M::process] read 67476 sequences (10000274 bp)...
[M::process] read 67236 sequences (10000039 bp)...
[M::mem_process_seqs] Processed 67476 reads in 1.895 CPU sec, 1.795 real sec
[M::process] read 67216 sequences (10000283 bp)...
[M::mem_process_seqs] Processed 67236 reads in 2.087 CPU sec, 1.946 real sec
[M::process] read 31399 sequences (4635971 bp)...
[M::mem_process_seqs] Processed 67216 reads in 2.439 CPU sec, 2.338 real sec
[M::mem_process_seqs] Processed 31399 reads in 1.275 CPU sec, 1.221 real sec
[main] Version: 0.7.17-r1188
[main] CMD: bwa mem data/ref.fasta data/SRR1705859.fastq
[main] Real time: 7.597 sec; CPU: 7.809 sec


In [12]:
! bwa mem data/ref.fasta data/SRR1705860.fastq | samtools view -S -b - | samtools sort - > data/sorted60.bam
! samtools index data/sorted60.bam

[M::bwa_idx_load_from_disk] read 0 ALT contigs
[M::process] read 67356 sequences (10000126 bp)...
[M::process] read 67208 sequences (10000122 bp)...
[M::mem_process_seqs] Processed 67356 reads in 1.817 CPU sec, 1.742 real sec
[M::process] read 67010 sequences (10000072 bp)...
[M::mem_process_seqs] Processed 67208 reads in 2.108 CPU sec, 1.986 real sec
[M::process] read 48390 sequences (7170166 bp)...
[M::mem_process_seqs] Processed 67010 reads in 2.175 CPU sec, 2.052 real sec
[M::mem_process_seqs] Processed 48390 reads in 1.624 CPU sec, 1.557 real sec
[main] Version: 0.7.17-r1188
[main] CMD: bwa mem data/ref.fasta data/SRR1705860.fastq
[main] Real time: 7.736 sec; CPU: 7.855 sec


In [13]:
! samtools mpileup -d 14000 -f data/ref.fasta data/sorted58.bam > data/58.mpileup
! samtools mpileup -d 14000 -f data/ref.fasta data/sorted59.bam > data/59.mpileup
! samtools mpileup -d 14000 -f data/ref.fasta data/sorted60.bam > data/60.mpileup

[mpileup] 1 samples in 1 input files
[mpileup] 1 samples in 1 input files
[mpileup] 1 samples in 1 input files


In [14]:
! java -jar VarScan.v2.4.4.jar mpileup2snp data/58.mpileup --min-var-freq 0.001 --variants --output-vcf 1 > data/VarScan_results58.vcf
! java -jar VarScan.v2.4.4.jar mpileup2snp data/59.mpileup --min-var-freq 0.001 --variants --output-vcf 1 > data/VarScan_results59.vcf
! java -jar VarScan.v2.4.4.jar mpileup2snp data/60.mpileup --min-var-freq 0.001 --variants --output-vcf 1 > data/VarScan_results60.vcf


Only SNPs will be reported
Min coverage:	8
Min reads2:	2
Min var freq:	0.001
Min avg qual:	15
P-value thresh:	0.01
Reading input from data/58.mpileup
1665 bases in pileup file
45 variant positions (45 SNP, 0 indel)
1 were failed by the strand-filter
44 variant positions reported (44 SNP, 0 indel)
Only SNPs will be reported
Min coverage:	8
Min reads2:	2
Min var freq:	0.001
Min avg qual:	15
P-value thresh:	0.01
Reading input from data/59.mpileup
1665 bases in pileup file
39 variant positions (39 SNP, 0 indel)
1 were failed by the strand-filter
38 variant positions reported (38 SNP, 0 indel)
Only SNPs will be reported
Min coverage:	8
Min reads2:	2
Min var freq:	0.001
Min avg qual:	15
P-value thresh:	0.01
Reading input from data/60.mpileup
1665 bases in pileup file
41 variant positions (41 SNP, 0 indel)
0 were failed by the strand-filter
41 variant positions reported (41 SNP, 0 indel)


In [25]:
! cat data/VarScan_results58.vcf | awk 'NR>24 {print $2, $4, $5, $10}'

38 T C 0/1:37:2740:2740:2722:18:0.66%:1.968E-4:35:31:2071:651:17:1
54 T C 0/1:24:5689:5658:5639:19:0.34%:3.2778E-3:35:32:4670:969:18:1
72 A G 0/1:26:5817:5814:5794:20:0.34%:2.0196E-3:38:37:4811:983:18:2
117 C T 0/1:31:7877:7848:7822:26:0.33%:6.5198E-4:37:36:6579:1243:23:3
165 T C 0/1:36:13756:13680:13640:39:0.29%:2.0227E-4:36:33:11366:2274:35:4
183 A G 0/1:48:13310:13214:13166:45:0.34%:1.4771E-5:36:34:10590:2576:39:6
218 A G 0/1:36:13715:13711:13671:39:0.28%:2.0227E-4:36:34:10651:3020:29:10
222 T C 0/1:21:13732:13673:13643:30:0.22%:6.8226E-3:37:35:10624:3019:27:3
235 T C 0/1:36:13624:13610:13569:39:0.29%:2.0226E-4:37:36:10513:3056:27:12
254 A G 0/1:23:13362:13315:13283:31:0.23%:4.7508E-3:36:35:10182:3101:28:3
276 A G 0/1:31:13702:13674:13636:36:0.26%:6.9461E-4:36:33:9684:3952:20:16
297 T C 0/1:20:13431:13360:13331:29:0.22%:9.7131E-3:36:34:9364:3967:20:9
340 T C 0/1:35:13613:13518:13477:38:0.28%:3.0697E-4:37:34:9060:4417:28:10
389 T C 0/1:23:13725:13620:13585:31:0.23%:4.7514E-3:37:37:84

In [26]:
! cat data/VarScan_results59.vcf | awk 'NR>24 {print $2, $4, $5, $10}'

44 T C 0/1:31:3809:3793:3775:18:0.47%:7.3433E-4:34:32:3006:769:17:1
158 A G 0/1:23:13173:13108:13075:31:0.24%:4.7503E-3:36:34:10258:2817:22:9
165 T C 0/1:35:13745:13650:13611:38:0.28%:3.07E-4:36:32:10950:2661:33:5
183 A G 0/1:23:13325:13209:13174:31:0.23%:4.7505E-3:36:34:10244:2930:28:3
193 A G 0/1:21:13213:13159:13129:30:0.23%:6.8212E-3:36:35:10017:3112:21:9
218 A G 0/1:36:13697:13697:13655:39:0.28%:2.0227E-4:36:35:10818:2837:31:8
222 T C 0/1:20:13739:13689:13660:29:0.21%:9.7142E-3:37:35:10806:2854:26:3
276 A G 0/1:21:13682:13645:13609:30:0.22%:6.8225E-3:37:34:9924:3685:20:10
319 T C 0/1:20:12969:12906:12875:28:0.22%:8.2532E-3:36:34:8883:3992:20:8
340 T C 0/1:23:13572:13471:13438:31:0.23%:4.7511E-3:37:36:9309:4129:25:6
356 A G 0/1:21:13530:13471:13440:30:0.22%:6.8221E-3:36:34:8721:4719:21:9
370 A G 0/1:21:13212:13140:13110:30:0.23%:6.8212E-3:37:36:7890:5220:13:17
398 A G 0/1:28:13771:13717:13682:34:0.25%:1.5315E-3:37:37:8129:5553:21:13
409 T C 0/1:20:13527:13423:13393:29:0.22%:9.7133E

In [27]:
! cat data/VarScan_results60.vcf | awk 'NR>24 {print $2, $4, $5, $10}'

38 T C 0/1:37:2573:2573:2554:18:0.7%:1.9651E-4:35:30:1827:727:17:1
44 T C 0/1:33:3822:3808:3789:19:0.5%:4.2089E-4:34:33:2963:826:18:1
105 A G 0/1:21:6356:6322:6301:19:0.3%:7.2691E-3:37:35:5066:1235:17:2
158 A G 0/1:38:12103:12056:12017:38:0.32%:1.5082E-4:36:35:9595:2422:33:5
165 T C 0/1:36:13660:13590:13549:39:0.29%:2.0225E-4:36:32:10404:3145:31:8
183 A G 0/1:23:13468:13372:13338:31:0.23%:4.7509E-3:36:36:9921:3417:19:12
193 A G 0/1:20:13283:13225:13196:29:0.22%:9.7126E-3:37:35:9696:3500:19:10
216 A G 0/1:20:13676:13674:13639:29:0.21%:9.7141E-3:37:36:9936:3703:18:11
235 T C 0/1:36:13639:13626:13586:39:0.29%:2.0226E-4:38:37:9871:3715:27:12
254 A G 0/1:24:13425:13365:13331:32:0.24%:3.2821E-3:36:35:9606:3725:22:10
271 A G 0/1:24:13733:13726:13694:32:0.23%:3.2827E-3:37:35:9257:4437:22:10
276 A G 0/1:54:13685:13659:13610:48:0.35%:3.7483E-6:37:35:9029:4581:32:16
297 T C 0/1:38:13506:13414:13373:40:0.3%:1.3247E-4:36:33:8856:4517:31:9
319 T C 0/1:20:12727:12667:12636:28:0.22%:8.2524E-3:36:35:83

As expected, there aro no variants with frequency above 1% in all 3 sequencings. 

In this table, all variants are sorted by unique position (`sort -u -n`)
| Position | Reference | Alternative | Frequency |
| --- | --- | --- | --- |
38 | T | C | 0.7%
44 | T | C | 0.47%
54 | T | C | 0.34%
72 | A | G | 0.34%
105 | A | G | 0.3%
117 | C | T | 0.33%
158 | A | G | 0.24%
165 | T | C | 0.28%
183 | A | G | 0.23%
193 | A | G | 0.22%
216 | A | G | 0.21%
218 | A | G | 0.28%
222 | T | C | 0.21%
235 | T | C | 0.29%
254 | A | G | 0.23%
271 | A | G | 0.23%
276 | A | G | 0.22%
297 | T | C | 0.22%
319 | T | C | 0.22%
340 | T | C | 0.23%
356 | A | G | 0.22%
370 | A | G | 0.23%
389 | T | C | 0.23%
398 | A | G | 0.25%
409 | T | C | 0.22%
414 | T | C | 0.22%
463 | A | G | 0.23%
494 | A | G | 0.22%
499 | A | G | 0.21%
516 | A | G | 0.22%
566 | A | G | 0.29%
595 | G | T | 0.35%
597 | A | G | 0.21%
660 | A | G | 0.25%
670 | A | G | 0.22%
722 | A | G | 0.22%
744 | A | G | 0.23%
759 | T | C | 0.23%
774 | T | C | 0.3%
793 | A | G | 0.24%
859 | A | G | 0.22%
898 | A | G | 0.22%
915 | T | C | 0.24%
1008 | T | G | 0.25%
1031 | A | G | 0.26%
1086 | A | G | 0.31%
1089 | A | G | 0.24%
1100 | T | C | 0.22%
1105 | A | G | 0.25%
1116 | A | G | 0.22%
1209 | A | G | 0.32%
1213 | A | G | 0.25%
1260 | A | C | 0.31%
1264 | T | C | 0.25%
1280 | T | C | 0.24%
1281 | T | C | 0.22%
1339 | T | C | 0.49%
1358 | A | G | 0.3%
1366 | A | G | 0.22%
1398 | T | C | 0.27%
1421 | A | G | 0.26%
1460 | A | G | 0.3%
1517 | A | G | 0.24%
1520 | T | C | 0.27%
1580 | T | C | 0.25%
1591 | T | C | 0.29%
1600 | T | C | 0.35%
1604 | T | C | 0.31%

Average frequency is **0.27%** and SD is **0.07%**, so probable mutation variants must have frequency above 0,45%


| Position | Reference | Alternative | Frequency | Residue change |
| --- | --- | --- | --- | --- |
| 307 | C | T | 0.92% |  P103S |
| 1458 | T | C | 0.86% | Y486Y |

Mutations in such positions are not present in control sequencing. 
One in 1458 position is silent, but one in 307 position is missense and 102th residue is in epitope D.

![alternative text](pic2.png)